Add GRANT query support for hive connector

amrutagokhale opened the pull request:
This is a prototype implementation of adding parsing support for GRANT query. I am listing the issues in the code that I am aware of:

- The current parsing code doesn't reflect the exact syntax of GRANT query that we want to support. I haven't iterated over the syntax, since this implementation is a prototype. I will refine it in due course. Here are some differences between the current syntax and the sql-99 standard syntax that we want to support:
-- Current syntax can accept only a single privilege at a time instead of a list of privileges
-- Current syntax doesn't take the keyword ALL which implies all privileges

- There are some TODO's in the code. These are issues to keep in mind; they may need to be fixed.

- I haven't removed all the comments from the code, specially in the hive connector code. Will do so eventually.

- The current implementation of function checkCanGrantTablePrivilege() performs the check of whether the current user can grant the privilege by checking whether the current user is the owner of the table and if so, the check passes. If not, the check fails. We may have to change this as per the security policy we decide to adopt (for ex., one of the privileges a user can have will be a 'GRANT' privilege that denotes whether a user can grant privileges to other users. We will then check whether the user has the GRANT privilege instead).

amrutagokhale commented:
@cawallin @rschlussel: can you please take a look?

cawallin commented:
Re your last bullet point: it should grant table privileges if, in the underlying data store, the user has the ability to grant privileges on a table.  Maybe I'm understanding it wrong, but in your implementation it might allow a user to do things it's not allowed to do.

amrutagokhale commented:
@cawallin Actually, the current implementation is more restrictive, because it only allows the owner of a table to grant privileges on that table to others. This is totally permissible, since in Hive, the owner does get the privilege to grant privileges, as noted here: https://cwiki.apache.org/confluence/display/Hive/SQL+Standard+Based+Hive+Authorization#SQLStandardBasedHiveAuthorization-Privileges -- "In the case of tables and views, the owner gets all the privileges with grant option."

cawallin commented:
@amrutagokhale But that's just for Hive, that's not an assumption you can make with databases in general. I bet that in PostgreSQL/MySQL you can revoke grant access to the owner of a table.

amrutagokhale commented:
@cawallin Point taken. The function checkCanGrantTablePrivilege() is only in the Hive connector code, as of now. Will keep this in mind, when I add code for other connectors.

rschlussel-zz commented:
I think even for hive we should check that the user has the grant privilege.  Both because we don't want to be more restrictive than hive.  But also because the owner having grant privileges is sort of incidental, and so we don't want the implementation to rely on that.

amrutagokhale commented:
Ok, will change this then in hive code too.

rschlussel-zz commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -886,6 +890,26 @@ private void invalidatePartitionCache(String databaseName, String tableName)
+            ImmutableList.Builder<HiveObjectPrivilege> privilegeBagList = ImmutableList.builder();
+            privilegeBagList.add(privilegeObject);
+
+            // Should you check whether the user already has the given privilege and if yes, return.
```
Is this a todo comment?  Also, is checking privileges that much faster than granting them?

amrutagokhale replied:
It's a todo. I am not entirely sure whether it is worth checking the privileges. I added the comment, just to keep that in mind.

amrutagokhale replied:
A faster way to check privileges would be to check the userTablePrivileges cache which stores the privileges. Before we do that, we may need to verify that the cache is populated, flushed and refreshed correctly. But this is definitely something to keep in mind.

rschlussel-zz commented on presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java:
```diff
@@ -177,6 +179,14 @@ public void checkCanSetCatalogSessionProperty(Identity identity, String property
+    @Override
+    public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege, SchemaTableName tableName)
+    {
+        if (!checkTablePermission(identity, tableName, OWNERSHIP)) {
```
Other users can also have grant privileges if they were granted permission 'WITH GRANT OPTION'

amrutagokhale replied:
yes, you're right. I am going to change OWNERSHIP -> GRANT

amrutagokhale replied:
done (in the new commits).

amrutagokhale commented:
@petroav @rschlussel can you please take a look at the new commits? I added 3 of them. From oldest to newest, the commits are:
1) Change the GRANT syntax to be able to specify multiple privileges in a single line. For ex., "grant insert, select to user on table_name". These changes are in the parser module.
2) Changes to the hive metastore to take into account the cache used to store user privileges. Main change in CachingHiveMetastore.java file is to return an object of type "RetryDriver" (an instance of RetryDriver class runs the code for a certain maximum number of attempts and times out after that)
3) Addresses the comments. This includes changing and renaming two classes:
PrestoIdentity -> IdentityNode and PrestoPrivilege -> PrivilegeNode. I also had to add a dependency in presto-parser module on presto-spi module, since the two classes in parser module now have members from the classes in SPI interface.

rschlussel-zz commented on presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java:
```diff
@@ -0,0 +1,74 @@
+
+            privileges.add(privilegeNode.getPrivilege());
+        }
+        // I don't fully understand the Identity object; what is user and what is principal? For now, I am passing grantee
```
I think this should be investigated more to ensure that it's done right.

amrutagokhale replied:
The "user" field corresponds to the user name and the "principal" field is the kerberos principal. @petroav pointed this out earlier. So passing the grantee's name as the first argument should be correct. With my new refactoring, this code to create a new identity has now moved to sql/tree/AstBuilder.java.

rschlussel-zz replied:
Can you just remove the comment then.

amrutagokhale replied:
Yup, will do.

rschlussel-zz commented:
The rest of the code looks good.  What's the plan for addressing the TODOs?

amrutagokhale commented:
@rschlussel All the TODOs are optimizations which may not be necessary to add right away. My plan is to leave the code as it is. I will clean up some of the old TODOs.

petroav commented:
I'm done reviewing. Have a look at my comments and let me know what you think. Before putting out a PR against FB Presto, I would re-organize the commits and give them better commit messages. Specifically the last refactor commit (72c17f197a9368ac61ff1c31c859699d5bf731a4) can be merged in with the commit that added the initial syntax (9ef7fa87114f70ea9790b9103eb5aa93512d9bcd). 3e1f37007f3ece3957a5e00fac786ae3c0ac2cd1 can also probably be merge in with 9ef7fa87114f70ea9790b9103eb5aa93512d9bcd. You'll have to use some `git rebase -i` magic to make it work but I think it's doable.

petroav commented:
I had a quick look at your replies. Things look good. I think after you re-shuffle the commits so that everything is cleaner, it would be beneficial if you posted another PR so we can take a look again. Right now I'm having trouble keeping the big picture in my head.

amrutagokhale commented:
Closing this old ticket.
